### DISCUSSION TITLE
Using pyinstaller with python-oracledb

### DISCUSSION THREAD

**Discussion by avmcohen** [oracle/python-oracledb#57] [Q&A] [answered]
Hello
Started a conversion process to move from cx_Oracle to oracledb (1.0.3).
I am using pyinstaller version 5.1 - works fine with cx_Oracle.
Python version : 3.9.5
I switched to oracledb, complied with pyinstaller, but get in runtime:
  _

> _Traceback (most recent call last):
>     File "dba_split.py", line 9, in <module>
>     File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
>     File "oracledb/__init__.py", line 38, in <module>
>     File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
>     File "oracledb/connection.py", line 39, in <module>
>     File "src/oracledb/impl/thin/crypto.pyx", line 32, in init oracledb.thin_impl
**>   ModuleNotFoundError: No module named 'cryptography'**
>   [15414] Failed to execute script '<my Python main script>' due to unhandled exception!_

Solutions suggested (add --paths, etc. have not helped)

_

**Comment by cjbj** [comment, marked as answer]
Does https://github.com/oracle/python-oracledb/issues/31 help?

**Reply by avmcohen** [reply]
I tried - didn't help. I added to pyinstaller:  
--hidden-import /usr/local/pyenv/versions/3.9.5/lib/python3.9/site-packages/cryptography/hazmat/primitives/kdf/

btw, python-oracledb works great with py (non-compiled) mode , imporoved more than 40% of data transfer performance !

**Reply by avmcohen** [reply]
Sorry, It worked finally by simply added in the code:
**_from cryptography.hazmat.primitives.kdf import pbkdf2_**

Also needed to add:
**_import encodings.idna_**

Thanks a lot !

**Reply by cjbj** [reply]
Thanks for sharing the steps.  And for the performance stat!

**Reply by ghost** [reply]
I am also getting the error below when using it with Lambda against Oracle DB 12c
python-oracledb thin mode cannot be used because "the cryptography package is not installed"

@cjbj  - are you saying all I have to add is the following two imports? do I need to install these via PIP ?
from cryptography.hazmat.primitives.kdf import pbkdf2
import encodings.idna

**Reply by cjbj** [reply]
@github4es you're asking the wrong person when it comes to Lambda!  Did you try what has been suggested here and in # 31 ?

**Comment by cjbj** [comment]
For future readers: with python-oracledb 1.1 when cryptography isn't configured, you are likely to get a new runtime error message  `DPY-3016: python-oracledb thin mode cannot be used because the cryptography package is not installed`

**Reply by avmcohen** [reply]
Hi
Note that Thin-mode is significantly faster in data upload/load. Is there a roadmap when the issue of case sensitive password will be solved in oracledb module ?

(Causing the ERROR:  _**oracledb.exceptions.NotSupportedError: DPY-3015: password verifier type 0x939 is not supported by python-oracledb in thin mode**_)

which force us to set: sec_case_sensitive_logon=TRUE , which is not always possible.

Thanks a lot.

**Reply by cjbj** [reply]
Thanks for the feedback about data uploads.

There are no plans to support the old "10g" verifier in Thin mode. Using Thick mode is the solution for legacy installations - until you can upgrade your database and/or password management practices !

**Reply by avmcohen** [reply]
Can you elaborate regarding "upgrade your database" ?.. I am working with Oracle 19.3. Will it be supported in later version ?

**Reply by cjbj** [reply]
It was a general comment about upgrading your environment and using case sensitive passwords, at which point you can use Thin mode.

The sec_case_sensitive_logon setting has been removed in 21c: https://docs.oracle.com/en/database/oracle/oracle-database/21/nfcon/security-solutions.html#GUID-FAF4C7A6-A2CD-4B9B-9A64-3705F693ECF0
